proof/refine/AARCH64/Invariants_H.thy

```diff
@@ -194,7 +210,10 @@ definition state_refs_of' :: "kernel_state \<Rightarrow> obj_ref \<Rightarrow> (
 primrec live0' :: "Structures_H.kernel_object \<Rightarrow> bool" where
   "live0' (KOTCB tcb) =
      (bound (tcbBoundNotification tcb) \<or>
-     (tcbState tcb \<noteq> Inactive \<and> tcbState tcb \<noteq> IdleThreadState) \<or>  tcbQueued tcb)"
+      tcbSchedPrev tcb \<noteq> None \<or>
+      tcbSchedNext tcb \<noteq> None \<or>
+      tcbState tcb \<noteq> Inactive \<and> tcbState tcb \<noteq> IdleThreadState \<or>
+      tcbQueued tcb)"
```

Comment:
I'm a bit surprised by this one: tcbQueued feels like it'll be linked to `tcbSched*`, but they're under an `∨`. Is the `tcbQueued` in Refine a flag only?

Reply:
`tcbQueued` appears only in the Haskell and C TCB objects. 

Maybe you already know this, but it took me a little while to think this through originally, so I'll write it out: it's possible for a ready queue to be a singleton, in which case the next and prev pointers are `None`, but the flag is still set. So they're closely related, but the flag can't be eliminated in favour of the next and prev pointers.

Reply:
You're absolutely right, and it did not occur to me immediately (mind too stuck in old proofs). Perhaps this is worthy of a comment above, especially as it took you a little while to think it through?

Reply:
I've also rearranged this a little to emphasise the connection between the next and prev fields and `tcbQueued` flag, and reformatted it a little to hopefully make the logical structure of it clearer.

Reply:
As for the comment, it's one of those annoying cases where a clear explanation ends up rather long and clunky. Here's what I have at the moment.

Note that for a TCB, we must consider its tcbSchedPrev and tcbSchedNext fields in addition to its
  tcbQueued field, since it may be sole member of a ready queue, in which case its tcbQueued field
  is @{term True} while its tcbSchedPrev and tcbSchedNext fields are both @{term None}.

Does that sound reasonable? I think it's horribly ugly.

Reply:
Does something like `For a TCB, we need to consider the next and previous pointers in addition to tcbQueued, as for a singleton queue they must both be None.` help make it shorter? The "scheduling" and "ready" stuff is probably not entirely exclusive; one day we might also get rid of IPC queue refinement, and I think that'd use the same pointers.

Reply:
I think your comment does at least look and sound much nicer. If others like it and think it’s helpful, then I will go with that. 

The other option would be to say something in the Haskell, where the `tcbQueued` field is first defined. We could say that the flag is still necessary in the case if singleton queues. 

But yes we certainly need both the flag for the ready queues and the flag for the release queue in MCS because they share the next and prev pointers, and we’ll see this all over again if we do the endpoint queues, I think. So it’s a general sort of thing.

Reply:
Oh and we do already have a new set of prev and next pointers for endpoint queues. `tcbEPnext` or something

Reply:
Oh wow, I did not realise that. In that case the "ready" / "sched" should make an appearance, maybe with `singleton ready queue`?. As for the Haskell, well, you know me, putting a short second comment there (as it's the spec) wouldn't be harmful.

Reply:
Nah, I think it's nicer to have these pointers as separate in the refinement, even if they (do they?) share them in the C. The TCB has lots of room, and our Haskell spec doesn't really exist on a real machine (the only reason we'd multiplex them is to save actual memory), so we're good.

Reply:
The more I think about it, the less I like a comment at all, anywhere, and the more I think that if we do have a comment, it should be in the spec. It’s really a justification for having the flag at all. Maybe I could say in the spec “membership in a ready queue cannot be inferred from the tcbSchedPrev and tcbSchedNext fields alone, since these will be None in the case of a singleton ready queue”. Or something more verbose and clunky, because it’s still not super clear

Reply:
If you needed to think about it and you wrote it, it's definitely worth a comment :-). I agree that the invariants are probably not the right place for it, so putting it in the spec would make sense. Having an explanation somewhere for how we expect the parts (flag + pointers) to play together would definitely be good.

Reply:
OK, comment now in the Haskell